lib/libzfs_core/libzfs_core.c

```diff
@@ -586,6 +589,114 @@ lzc_get_holds(const char *snapname, nvlist_t **holdsp)
 	return (lzc_ioctl(ZFS_IOC_GET_HOLDS, snapname, NULL, holdsp));
 }
 
+static unsigned int
+max_pipe_buffer(int infd)
+{
+#if __linux__
+	static unsigned int max;
+	if (max == 0) {
+		max = 1048576; /* fs/pipe.c default */
+
+		FILE *procf = fopen("/proc/sys/fs/pipe-max-size", "re");
+		if (procf != NULL) {
+			if (fscanf(procf, "%u", &max) <= 0) {
+				/* errors ignored: max untouched if parse failed */
+			}
+			fclose(procf);
+		}
+	}
+
+	unsigned int cur = fcntl(infd, F_GETPIPE_SZ);
+	if (cur < max && fcntl(infd, F_SETPIPE_SZ, max) != -1)
+		cur = max;
+	return (cur);
+#else
+	/* FreeBSD automatically resizes */
+	(void) infd;
+	return (BIG_PIPE_SIZE);
+#endif
+}
+
+#if __linux__
+struct send_worker_ctx {
+	int from;	/* read end of pipe, with send data; closed on exit */
+	int to;		/* original arbitrary output fd; mustn't be a pipe */
+};
+
+static void *
+send_worker(void *arg)
+{
+	struct send_worker_ctx *ctx = arg;
+	unsigned int bufsiz = max_pipe_buffer(ctx->from);
+	ssize_t rd;
+
+	while ((rd = splice(ctx->from, NULL, ctx->to, NULL, bufsiz,
+	    SPLICE_F_MOVE | SPLICE_F_MORE)) > 0)
+		;
+
+	int err = (rd == -1) ? errno : 0;
+	close(ctx->from);
+	return ((void *)(uintptr_t)err);
+}
+#endif
+
+/*
+ * Since Linux 5.10, 4d03e3cc59828c82ee89ea6e27a2f3cdf95aaadf
+ * ("fs: don't allow kernel reads and writes without iter ops"),
+ * ZFS_IOC_SEND* will EINVAL when writing to /dev/null, /dev/zero, &c.
+ *
+ * This wrapper transparently executes func() with a pipe
+ * by spawning a thread to copy from that pipe to the original output
+ * in the background.
+ *
+ * Returns the error from func(), if nonzero,
+ * otherwise the error from the thread.
+ *
+ * No-op if orig_fd is -1, already a pipe (but the buffer size is bumped),
+ * and on not-Linux; as such, it is safe to wrap/call wrapped functions
+ * in a wrapped context.
+ */
+int
+lzc_send_wrapper(int (*func)(int, void *), int orig_fd, void *data)
+{
+#if __linux__
+	struct stat sb;
+	if (orig_fd != -1 && fstat(orig_fd, &sb) == -1)
+		return (errno);
+	if (orig_fd == -1 || S_ISFIFO(sb.st_mode)) {
+		if (orig_fd != -1)
+			(void) max_pipe_buffer(orig_fd);
+		return (func(orig_fd, data));
+	}
+
+	int rw[2];
+	if (pipe2(rw, O_CLOEXEC) == -1)
+		return (errno);
+
+	int err;
+	pthread_t send_thread;
+	struct send_worker_ctx ctx = {.from = rw[0], .to = orig_fd};
+	if ((err = pthread_create(&send_thread, NULL, send_worker, &ctx))
```

Comment:
It.. doesn't. But neither does inverting it (well, it does make it a bit better in the trivialest case (writing to file), but it's actively worse because the ioctl gets EINTRed and the send fails; as-is you can resume it manually and it *will* continue to work in my testing).

As an added bonus, 2.1.2-1.1 is also victim to this, so this is not a regression here.

Reply:
Fascinating. I really envy Illumos their extremely simple handling of this problem (`issig` just handles SIGSTOP and all the job control stuff internally). Once this lands I'll spend some time doing experiments and see if I can find an approach that works reliably for Linux, thanks for giving it a spin.